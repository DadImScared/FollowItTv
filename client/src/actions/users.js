
import axios from 'axios';


export const registerUser = async (payload) => {
  return await axios.post('/api/v1/register/', payload);
};

export const reSendEmailConfirm = async (email) => {
  return await axios.post('/resend_email', { email });
};
