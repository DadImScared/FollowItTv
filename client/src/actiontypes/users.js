
export const LOG_IN = 'users/LOG_IN';
export const LOG_OUT = 'users/LOG_OUT';
