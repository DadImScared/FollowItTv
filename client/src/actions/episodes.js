
import { ADD_EPISODES } from '../actiontypes/episodes';

export const addEpisodes = (episodes) => {
  return {
    type: ADD_EPISODES,
    episodes
  };
};
