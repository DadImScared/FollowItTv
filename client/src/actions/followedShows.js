
import _ from 'lodash';
import axios from 'axios';

import axiosOptions from './axiosOptions';
import * as followedActionTypes from '../actiontypes/followedShows';
import { addShows } from './shows';

export const addDay = (day, shows) => {
  return {
    type: followedActionTypes.ADD_DAY,
    day,
    shows
  };
};

export const followShow = (days, id) => {
  return {
    type: followedActionTypes.FOLLOW_SHOW,
    days,
    id
  };
};

export const unfollowShow = (days, id) => {
  return {
    type: followedActionTypes.UNFOLLOW_SHOW,
    days,
    id
  };
};

export const postFollow = async (showId) => {
  return await axios.post(`/api/v1/followshow/${showId}/`, {}, axiosOptions());
};

export const addShowIds = (showIdList) => {
  return {
    type: followedActionTypes.ADD_SHOW_IDS,
    showIdList
  };
};

export const extractShowData = (data) => {
  let isArrayOfNumbers;
  const showIds = new Set();
  const showData = {};
  if (typeof data[0] === 'number') {
    isArrayOfNumbers = true;
  }
  else {
    data.forEach((show) => {
      showIds.add(show.id);
      showData[show.id] = show;
    });
  }
  return {
    showData,
    showIds: isArrayOfNumbers ? data:[...showIds]
  };
};

export const getFollowedShows = async (dispatch, day = '') => {
  try {
    const { data } = await axios.get(`/api/v1/followshows/${day ? `${_.capitalize(day)}/`:''}`, axiosOptions());
    const { showData, showIds } = extractShowData(data);
    if (day) {
      dispatch(addShows(showData));
    }
    dispatch(addShowIds(showIds));
  }
  catch (e) {
    throw e;
  }
};
