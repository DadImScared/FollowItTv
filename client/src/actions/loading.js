
import { LOADING_START, LOADING_END } from '../actiontypes/loading';

export const loadingStart = (requestId) => {
  return {
    type: LOADING_START,
    requestId
  };
};

export const loadingEnd = (requestId) => {
  return {
    type: LOADING_END,
    requestId
  };
};
