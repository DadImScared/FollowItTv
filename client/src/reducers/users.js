
import Cookies from 'js-cookie';

import { LOG_IN, LOG_OUT } from '../actiontypes/users';


const initialState = {
  loggedIn: Cookies.get('token') !== undefined,
  token: Cookies.get('token')
};

function users(state=initialState, action) {
  switch(action.type) {
  case LOG_IN:
    return {
      loggedIn: true,
      token: action.token
    };
  case LOG_OUT:
    return {
      loggedIn: false,
      token: ''
    };
  default:
    return state;
  }
}

export default users;
