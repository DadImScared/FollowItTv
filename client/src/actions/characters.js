
import { ADD_CHARACTERS } from '../actiontypes/characters';
import { loadingEnd, loadingStart } from './loading';
import { castSchema } from '../schemas/person';
import { normalize } from 'normalizr';
import { baseUrl } from './seasons';
import axios from 'axios';

export const addCharacters = (entities) => ({
  type: ADD_CHARACTERS,
  ...entities
});

const getCast = async (showId) => {
  return await axios.get(`${baseUrl}shows/${showId}/cast`);
};

export const handleCastData = (data) => {
  const { entities: { person, character }, result } = normalize(data, castSchema);
  const characterIds = result.reduce((acc, { character }) => {
    const newAcc = [...acc];
    if (!newAcc.includes(character)) {
      newAcc.push(character);
    }
    return newAcc;
  }, []);
  return {
    person,
    character,
    result,
    characterIds
  };
};

export const requestCast = (showId) => {
  return async (dispatch) => {
    const loadingId = `GET_CAST_${showId}`;
    dispatch(loadingStart(loadingId));
    try {
      const { data } = await getCast(showId);
      const { person: people, character: characters, characterIds } = handleCastData(data);
      dispatch(addCharacters({ id: showId, people, characters, characterIds }));
      dispatch(loadingEnd(loadingId));
    }
    catch (e) {
      dispatch(loadingEnd(loadingId));
      throw e;
    }
  };
};
