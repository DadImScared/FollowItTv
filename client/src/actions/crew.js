
import axios from 'axios';
import _ from 'lodash';
import { normalize } from 'normalizr';

import { ADD_CREW } from '../actiontypes/crew';
import { loadingStart, loadingEnd } from './loading';
import { baseUrl } from './seasons';
import { crewSchema } from '../schemas/person';

export const addCrew = (entities) => ({
  type: ADD_CREW,
  ...entities
});

const getCrew = async (showId) => await axios.get(`${baseUrl}shows/${showId}/crew`);

export const handleCrewData = (data) => {
  const { entities, result } = normalize(data, crewSchema);
  const crewMembers = result.reduce((acc, crewMember) => {
    return {
      ...acc,
      [crewMember.person]: _.union(_.get(acc, crewMember.person, []), [ crewMember.type ])
    };
  }, {});
  return {
    ...entities,
    result,
    crewMembers
  };
};

export const requestCrew = (showId) => {
  return async (dispatch) => {
    const requestId = `GET_CREW_${showId}`;
    dispatch(loadingStart(requestId));
    try {
      const { data } = await getCrew(showId);
      const { person: people, crewMembers } = handleCrewData(data);
      dispatch(addCrew({ id: showId, people, crewMembers }));
      dispatch(loadingEnd(requestId));
    }
    catch(e) {
      dispatch(loadingEnd(requestId));
      throw e;
    }
  };
};
