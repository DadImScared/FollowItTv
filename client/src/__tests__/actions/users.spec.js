
import axios from 'axios';

import { LOG_OUT, LOG_IN } from '../../actiontypes/users';
import { registerUser, reSendEmailConfirm, logIn, logOut } from '../../actions/users';

jest.mock('axios');


const payload = {
  email: 'thing@mail.com',
  password1: 'password',
  password2: 'password'
};

const success = {
  data: 'success'
};

describe('user actions', () => {
  describe('registerUser', () => {
    it('should return success on valid payload', async () => {
      axios.post.mockImplementationOnce(() => success);
      const { data } = await registerUser(payload);

      expect(data).toEqual('success');
    });

    it('should throw error on invalid payload', async () => {
      axios.post.mockImplementationOnce(() => {
        throw {
          response: {
            data: 'error'
          }
        };
      });
      try {
        await registerUser({});
      }
      catch ({ response: { data } }) {
        expect(data).toEqual('error');
      }
    });
  });

  describe('logIn', () => {
    it('should create logIn action', () => {
      const expectedAction = {
        type: LOG_IN,
        token: 'token'
      };
      expect(logIn('token')).toEqual(expectedAction);
    });
  });

  describe('logOut', () => {
    it('should create logOut action', () => {
      const expectedAction = {
        type: LOG_OUT
      };
      expect(logOut()).toEqual(expectedAction);
    });
  });
});
