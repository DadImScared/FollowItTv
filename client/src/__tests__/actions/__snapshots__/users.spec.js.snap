// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user actions logInUser successful login 1`] = `
Array [
  Object {
    "token": "token here",
    "type": "users/LOG_IN",
  },
]
`;
